cudax/include/cuda/experimental/__stf/internal/scalar_interface.cuh

```diff
@@ -0,0 +1,275 @@
+  size_t size() const
+  {
+    return sizeof(T);
+  }
+};
+
+/**
+ * @brief Data interface to manipulate the void interface in the CUDA stream backend
+ */
+template <typename T>
+class scalar_stream_interface : public stream_data_interface_simple<scalar<T>>
+{
+public:
+  using base = stream_data_interface_simple<scalar<T>>;
+  using typename base::shape_t;
+
+  scalar_stream_interface(scalar<T> val)
+      : base(::std::move(val))
+  {}
+  scalar_stream_interface(typename base::shape_t s)
+      : base(s)
+  {}
+
+  /// Copy the content of an instance to another instance : this is a no-op
+  void stream_data_copy(
+    const data_place& dst_memory_node,
+    instance_id_t dst_instance_id,
+    const data_place& src_memory_node,
+    instance_id_t src_instance_id,
+    cudaStream_t stream) override
+  {
+    assert(src_memory_node != dst_memory_node);
+
+    cudaMemcpyKind kind = cudaMemcpyDeviceToDevice;
+    if (src_memory_node == data_place::host)
+    {
+      kind = cudaMemcpyHostToDevice;
+    }
+
+    if (dst_memory_node == data_place::host)
+    {
+      kind = cudaMemcpyDeviceToHost;
+    }
+
+    const scalar<T>& src_instance = this->instance(src_instance_id);
+    const scalar<T>& dst_instance = this->instance(dst_instance_id);
+
+    size_t sz = sizeof(T);
+
+    cuda_safe_call(cudaMemcpyAsync((void*) dst_instance.addr, (void*) src_instance.addr, sz, kind, stream));
+  }
+
+  void stream_data_allocate(
+    backend_ctx_untyped&,
+    const data_place& memory_node,
+    instance_id_t instance_id,
+    ::std::ptrdiff_t& s,
+    void**,
+    cudaStream_t stream) override
+  {
+    scalar<T>& instance = this->instance(instance_id);
+    T* base_ptr;
+
+    if (memory_node == data_place::host)
+    {
+      // Fallback to a synchronous method as there is no asynchronous host allocation API
+      cuda_safe_call(cudaStreamSynchronize(stream));
+      cuda_safe_call(cudaHostAlloc(&base_ptr, sizeof(T), cudaHostAllocMapped));
+    }
+    else
+    {
+      cuda_safe_call(cudaMallocAsync(&base_ptr, sizeof(T), stream));
+    }
+
+    instance.addr = base_ptr;
+
+    s = sizeof(T);
+  }
+
+  /// Pretend we deallocate an instance (no-op)
+  void stream_data_deallocate(backend_ctx_untyped&, const data_place&, instance_id_t, void*, cudaStream_t) override {}
+
+  bool pin_host_memory(instance_id_t) override
+  {
+    // no-op
+    return false;
+  }
+
+  void unpin_host_memory(instance_id_t) override {}
+};
+
+/**
+ * @brief Define how the CUDA stream backend must manipulate this scalar<T> interface
+ *
+ * Note that we specialize cuda::experimental::stf::shape_of to avoid ambiguous specialization
+ *
+ * @extends streamed_interface_of
+ */
+template <typename T>
+struct streamed_interface_of<scalar<T>>
+{
+  using type = scalar_stream_interface<T>;
+};
+
+/**
+ * @brief Data interface to manipulate the void interface in the CUDA graph backend
+ */
+template <typename T>
+class scalar_graph_interface : public graph_data_interface<scalar<T>>
```

Comment:
Since this is not really implemented, we need to add calls to abort() so that we don't think it is implemented by mistake